feat: new outlier detection ui (dhis2-10259)

HendrikThePendric opened the pull request:
Some notes:
- Previously this app had a "Min-max outlier analysis" and "Std Dev outlier analysis" section
- The new "Outlier detection" section replaces both, and has been implemented by editing the "Std Dev outlier analysis" section
- Currently the solution is lacking the functionality to "mark outliers for follow-up". This will be introduced later on, because more backend work is required for this.
- I've adjusted some failing tests but didn't really add thorough test coverage based on the new UI. I chose not to do this because this app is in dire need of being completely rewritten in the near future.
- Along the same lines: I tried to implement this following the architecture / coding practices that were used in the app already. It's not really idiomatic react code, but it works.

ismay commented on src/components/available-organisation-units-tree/AvailableOrganisationUnitsTree.js:
```diff
@@ -64,18 +68,29 @@ class AvailableOrganisationUnitsTree extends PureComponent {
+        const { path } = orgUnit
+        const paths = selected.includes(path)
+            ? selected.filter(selectedPath => selectedPath !== path)
+            : [...this.state.selected, orgUnit.path]
```
I assume you can just use path here instead of orgUnit.path

amcgee replied:
Yep, agreed @ismay

ismay commented on src/components/download-as/DownloadAs.js:
```diff
@@ -10,40 +10,46 @@ class DownloadAs extends PureComponent {
+    getHrefs() {
+        const { baseUrl } = this.context.d2.Api.getApi()
+        const { endpoint, fileTypes } = this.props
         const timestamp = Date.now()
+        const queryStr = this.props.queryStr || `t=${timestamp}`
```
You explained this during our call, but maybe for posterity it'd be nice to leave a short comment here explaining what's going on with the timestamp, and why we're falling back to that, etc.

amcgee replied:
Is this for cache busting?

ismay replied:
Yeah that's the only use I can think of, but I'm not sure.

HendrikThePendric replied:
I looked into this some more. I do believe this is some sort of cache buster, because if I access the url without the timestamp it still works. Since it was already present I am reluctant to remove it.

ismay commented on src/components/outlier-analysis-table/OutlierAnalysisTable.js:
```diff
@@ -118,37 +174,46 @@ OutlierAnalyisTable.generateElementKey = e =>
     }-${e.dataElementId}`
 
 OutlierAnalyisTable.convertElementFromApiResponse = e => ({
```
So from what I can gather this method is used in `src/pages/outlier-detection/OutlierDetection.js`. Not something you can do anything about, but seems so confusing that this is defined here as a static method. But yeah, guess it fits with the other patterns used in this app.

ismay reviewed (commented):
Looks good 👍. It's not the easiest app to review, but it helped that you walked me through it in the call. Let me know when you've had time to take a look at the things you mentioned during our call and I can take another look.

By the way, I think that this app would probably be a good candidate for some higher prioritization on the 'refactor to the app platform' list. Seems like it would really benefit from it.

amcgee commented on src/pages/outlier-detection/OutlierDetection.test.js:
```diff
@@ -105,8 +104,16 @@ describe('Test <StdDevOutlierAnalysis /> rendering:', () => {
 
-    it('Renders a "SelectField" to choose Standard Deviation.', () => {
-        expect(wrapper.find(SelectField).length).toBe(1)
+    it('Renders a an input to choose to choose Algorithm.', () => {
+        expect(wrapper.find('#algorithm').length).toBe(1)
+    })
+
+    it('Renders a an input to choose to choose Threshold.', () => {
+        expect(wrapper.find('#threshold').length).toBe(1)
+    })
+
+    it('Renders a an input to choose to choose Max Results.', () => {
+        expect(wrapper.find('#max-results').length).toBe(1)
```
```suggestion
    it('Renders an input to choose Algorithm.', () => {
        expect(wrapper.find('#algorithm').length).toBe(1)
    })

    it('Renders an input to choose Threshold.', () => {
        expect(wrapper.find('#threshold').length).toBe(1)
    })

    it('Renders an input to choose Max Results.', () => {
        expect(wrapper.find('#max-results').length).toBe(1)
```

amcgee commented on src/components/outlier-analysis-table/OutlierAnalysisTable.js:
```diff
@@ -15,51 +15,88 @@ import DownloadAs from '../../components/download-as/DownloadAs'
 import { apiConf } from '../../server.conf'
 import cssPageStyles from '../../pages/Page.module.css'
-import jsPageStyles from '../../pages/PageStyles'
+// import jsPageStyles from '../../pages/PageStyles'
```
Should we just remove this line or is there a reason to leave it as a comment?  Same with several other lines below

ismay replied:
I should have explained what Hendrik and I talked about in the call actually. So Hendrik will move all the commented code to a separate branch, it's functionality that he created but that won't work with the current backend (if I'm not mistaken). In the future we should be able to merge that as well, once the functionality is there.

HendrikThePendric replied:
As discussed with @ismay I will leave the comments in, to avoid merge conflicts later. They will be uncommented fairly soon.

amcgee reviewed (commented):
As @ismay mentioned it's a bit tough to see what's going on here... I made a few super minor comments.  If you'd like me to review the code more thoroughly let me know, otherwise I'll defer to @ismay since he seems to have some context.

I'll try running this locally to test.

amcgee commented:
App seems to work well, I'm trying to get access to a WHO database (https://who-demos.dhis2.org/dq/dhis-web-dashboard/#/) so we can spin up a 2.35 server and use that as a test backend.

One thing I noticed when viewing outliers - it's not really clear what the "outlier" is... should we highlight the "Value" column with a color or boldface and/or add some indicators to the other columns to reflect what they represent?  @cooper-joe @larshelge 

![Screen Shot 2021-01-25 at 23 11 14](https://user-images.githubusercontent.com/947888/105772741-f06c5f80-5f62-11eb-8b6c-b24d9f020997.png)

For example - in this screenshot, the "value" column is between "min" and "max", but "mean" is not.  There are two different "types" of columns here, the columns that represent the "expected" or "normal" range (min, max, mean, std dev) and the columns that represent this particular outlier (value, deviation, and z-score).  So maybe those two sets of columns should be grouped together?  We should try, if possible, to convey visibly that the "Z-Score" is the WEIGHT of how much this particular value is an outlier, and should also sort by descending z-score

For reference, I believe the formulas here are `deviation = value - mean` (it's possible deviation could be negative, I think...) and `deviation = stddev * z-score` (more or less, there might be some more going on under the hood).  But basically these are all saying the same things... value is obviously important, but z-score tells us just how "wrong" that value is.

I think it's fine to merge this as-is for now, but I'd like to explore how we can make this more user-friendly in the future (maybe when refactoring the app and/or adding the follow-up workflow)

amcgee commented:
Another point of feedback (minor): when running outlier detection it will navigate to a new "page" with a full-screen table.  However, the route in `window.location` doesn't change.  As such, clicking the "back arrow" on the page itself will return you to the expected page (the outlier detection parameters page, with state maintained so you don't need to re-enter any parameters) however using the back button in the browser will return you to the start screen.  This also clears the state, so everything is reset if you then navigate manually to the outlier detection page again.

Would it be straightforward to push a new route onto history here so that we can also use the native back button?  If it's a lot of work I don't think this is worth much effort, but if we can fix it easily we should.

cooper-joe commented:
> I think it's fine to merge this as-is for now, but I'd like to explore how we can make this more user-friendly in the future (maybe when refactoring the app and/or adding the follow-up workflow)

Just a note to say I agree with @amcgee - I think we could do some helpful things here with the way we display outliers. I'd certainly be interested in taking a look, so let me know if this gets on the roadmap.

ismay commented on src/components/formatters/FormattedNumber.js:
```diff
@@ -1,17 +1,28 @@
 )
 
+FormattedNumber.defaultProps = {
+    minimumFractionDigits: 0,
```
Hendrik and I talked about this as well. Seems like the default has changed here. Just commenting so other reviewers know that this was discussed already.

ismay commented on src/pages/outlier-detection/OutlierDetection.test.js:
```diff
@@ -132,17 +131,20 @@ describe('Test <StdDevOutlierAnalysis /> rendering:', () => {
         })
         expect(wrapper.find(IconButton)).toHaveLength(1)
         expect(wrapper.find(IconButton).props().style.display).toBe('inline')
+        console.log(wrapper.find(OutlierAnalyisTable))
```
FYI, we discussed this as well. Will be removed.

HendrikThePendric commented:
With regards to https://github.com/dhis2/data-quality-app/pull/662#issuecomment-767152213 by @amcgee . I think that re-arranging the column order would make sense in the current PR. We are indeed showing values referring to the the dataset (or subset when using data-start|end-date), and values referring to the data-value and we are mixing the columns. So grouping the columns using a data-value vs data-set approach _could_ make sense..... That being said, when reviewing outliers the data-analyst will be mainly comparing these things: Value VS Mean (perhaps also taking min and max into account) and Deviation VS StdDeviation. On balance, I think I would still prefer keeping the grouping the data in a DataValue VS DataSet way, so probably the column order I'd suggest would be:

1. First the Data Value, so `Value`, `ZScore` and `Deviation`
2. Then the Data Set, so `StDeviation` (which is now also conveniently directly after `Deviation`) `Mean`, `Min` , `Max`

@amcgee @larshelge do you agree?

HendrikThePendric commented:
> Would it be straightforward to push a new route onto history here so that we can also use the native back button? If it's a lot of work I don't think this is worth much effort, but if we can fix it easily we should.

I think I'd like to leave that for when we refactor things. This apps a bit quirky and it wouldn't be as easy to fix as you might hope/expect.

amcgee commented:
> First the Data Value, so Value, ZScore and Deviation
> Then the Data Set, so StDeviation (which is now also conveniently directly after Deviation) Mean, Min , Max

Makes sense to me

HendrikThePendric commented:
I've addressed all points discussed in the reviews and meetings, apart from the commented code which is going to get uncommented in the near future. So ready for re-review.

ismay reviewed (approved):
I think the points I mentioned have all been addressed. I think that just leaves @amcgee and @cooper-joe's points, I'll leave it to them to take a look at whether those are all now resolved as well.

HendrikThePendric commented:
> I think that just leaves @amcgee and @cooper-joe's points, I'll leave it to them to take a look at whether those are all now resolved as well.

I'm going to assume they have been since I have rearranged the columns according to my suggestion that received a 👍 . I also know that people would like to start testing this out, so I am going to merge this in now.
